core/src/main/java/com/datastrato/gravitino/authorization/PermissionManager.java

```diff
@@ -0,0 +1,319 @@
+import com.datastrato.gravitino.meta.AuditInfo;
+import com.datastrato.gravitino.meta.GroupEntity;
+import com.datastrato.gravitino.meta.RoleEntity;
+import com.datastrato.gravitino.meta.UserEntity;
+import com.datastrato.gravitino.utils.PrincipalUtils;
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * PermissionManager is used for managing the logic the granting and revoking roles. Role is used
+ * for manging permissions. GrantManager will filter the invalid roles, too.
+ */
+class PermissionManager {
+  private static final Logger LOG = LoggerFactory.getLogger(PermissionManager.class);
+
+  private final EntityStore store;
+
+  PermissionManager(EntityStore store) {
+    this.store = store;
+  }
+
+  public boolean grantRoleToUser(String metalake, String role, String user) {
+    try {
+      AuthorizationUtils.checkMetalakeExists(store, metalake);
+
+      checkRoleExists(metalake, role);
+
+      RoleEntity roleEntity =
+          store.get(
+              NameIdentifierUtils.ofRole(metalake, role), Entity.EntityType.ROLE, RoleEntity.class);
+
+      store.update(
+          NameIdentifierUtils.ofUser(metalake, user),
+          UserEntity.class,
+          Entity.EntityType.USER,
+          userEntity -> {
+            List<RoleEntity> roleEntities = removeInvalidRoles(metalake, userEntity.roles());
+            List<String> roleNames = Lists.newArrayList(toRoleNames(roleEntities));
+            List<Long> roleIds = Lists.newArrayList(toRoleIds(roleEntities));
+
+            if (roleNames.contains(roleEntity.name())) {
+              throw new RoleAlreadyExistsException(
+                  "Role %s already exists in the user %s of the metalake %s", role, user, metalake);
+            }
+
+            roleNames.add(roleEntity.name());
+            roleIds.add(roleEntity.id());
+            AuditInfo auditInfo =
+                AuditInfo.builder()
+                    .withCreator(userEntity.auditInfo().creator())
+                    .withCreateTime(userEntity.auditInfo().createTime())
+                    .withLastModifier(PrincipalUtils.getCurrentPrincipal().getName())
+                    .withLastModifiedTime(Instant.now())
+                    .build();
+
+            return UserEntity.builder()
+                .withNamespace(userEntity.namespace())
+                .withId(userEntity.id())
+                .withName(userEntity.name())
+                .withRoleNames(roleNames)
+                .withRoleIds(roleIds)
+                .withAuditInfo(auditInfo)
+                .build();
+          });
+      return true;
+    } catch (NoSuchEntityException nse) {
+      LOG.warn("User {} does not exist in the metalake {}", user, metalake, nse);
```

Comment:
How do you distinguish this nse is from get role operation or update user operation?

Reply:
I checked the role whether exists before this.

Reply:
I misunderstand your meaning. I will correct this.